Project Management: Prompt user to link GitHub account to WordPress.org profile

aduth opened the pull request:
Related:

- #21013
- https://github.com/WordPress/gutenberg/projects/24#card-34839803
- https://meta.trac.wordpress.org/ticket/4447
- https://make.wordpress.org/core/2020/03/19/associating-github-accounts-with-wordpress-org-profiles/

This pull request seeks to fix and enhance the existing "First Time Contributor" project management automation to prompt the user to link their GitHub account to their WordPress.org profile if necessary for release props credit.

Currently, the prompt is:

> Congratulations on your first merged pull request! We'd like to credit you for your contribution in the post announcing the next WordPress release, but we can't find a WordPress.org profile associated with your GitHub account. When you have a moment, visit the following URL and click "link your GitHub account" under "GitHub Username" to link your accounts:
> 
> https://profiles.wordpress.org/me/profile/edit/
> 
> And if you don't have a WordPress.org account, you can create one on this page:
> 
> https://login.wordpress.org/register
>
> Kudos!

<details><summary>(Click to expand for initial prompt proposed)</summary>
>Congratulations on your first merged pull request! We would like to give you credit for your contribution in the next WordPress release, but we were unable to find a WordPress.org profile associated with your GitHub account. At your convenience, please visit the following URL and click "link your GitHub account" under "GitHub Username" to initiate the process to link your accounts:
>
>https://profiles.wordpress.org/me/profile/edit/
>
>If you do not have a WordPress.org account, you can create one at the following page:
>
>https://login.wordpress.org/register
</details>

~Feedback is very welcome for adjustments to the phrasing (I've also applied "Needs Copy Review" label to this pull request).~ Reviewed at https://github.com/WordPress/gutenberg/pull/21221#issuecomment-605945989.

The prompt is presented under the following conditions:

- This is the first commit merged for the user.
- A WordPress.org user profile is not known to be associated with their GitHub account.

As noted, this also _fixes_ the "First Time Contributor" task, which was intended to be applying [the "First Time Contributor" label](https://github.com/WordPress/gutenberg/issues?q=is%3Aopen+is%3Aissue+label%3A%22First-time+Contributor%22) to a pull request opened by a new contributor. Unfortunately, due to [permission constraints](https://help.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token#permissions-for-the-github_token) of GitHub Actions, labels cannot be applied by actions if the originating repository is a fork of Gutenberg. Since the vast majority of new contributors will not have branching permissions of this repository, this made the task near-useless. This is resolved by adopting a similar approach of #20058 : The task is run on _commit to master_, where full permissions are granted. This means that the label is only applied after merge, which is not perfect, but certainly an improvement. Coincidentally, this complements the goal of prompting the user to link their accounts upon merge.

**Testing Instructions:**

There are unit tests:

```
npm run test-unit packages/project-management-automation/lib/test/add-first-time-contributor-label.js
```

But the most accurate test won't be 'til we can have it running live 😅

aduth committed:
Project Management Automation: Add getAssociatedPullRequest utility

aduth committed:
Project Management Automation: Add First-Time Contributor label on push

aduth committed:
Project Management Automation: Include as dependency in root package

aduth committed:
Project Management Automation: Prompt user to link GitHub account to profile

aduth committed:
Project Management Automation: Update documentation for first contribution prompt

aduth committed:
Project Management Automation: Check for single commit of author

dd32 commented:
The API usage looks correct, however, the API will throw a `404` response when the user isn't known.
From what I can tell, `got()` will probably consider that an error. I think you need to set the [throwhttperrors](https://www.npmjs.com/package/got#throwhttperrors) option? But that's just from glancing at the docs.

Please also make sure the request comes with a clear UserAgent that our systems team can differentiate from other random clients to avoid unintentional blocking/rate-limiting.

michelleweber commented:
This is great! I'd use a few contractions -- they'll make it sound a little friendlier/less formal -- and tighten up a sentence here and there, but nothing big:

> Congratulations on your first merged pull request! We'd like credit you for your contribution in the next WordPress release, but we can't find a WordPress.org profile associated with your GitHub account. When you have a moment, visit the following URL and click "link your GitHub account" under "GitHub Username" to link your accounts:
> 
> https://profiles.wordpress.org/me/profile/edit/
> 
> And if you don't have a WordPress.org account, you can create one on this page:
> 
> https://login.wordpress.org/register
>
> Kudos!

desrosj commented on packages/project-management-automation/lib/add-first-time-contributor-label.js:
```diff
@@ -1,50 +1,131 @@
-	const author = payload.pull_request.user.login;
+	if ( payload.ref !== 'refs/heads/master' ) {
+		debug(
+			'add-first-time-contributor-label: Commit is not to `master`. Aborting'
```
Would it make sense to display the prompt regardless of branch? If they are opening a PR, it shows they are interested in contributing. Why not catch them the first time regardless of branch?

aduth replied:
>Would it make sense to display the prompt regardless of branch? If they are opening a PR, it shows they are interested in contributing. Why not catch them the first time regardless of branch?

Unfortunately, it's not possible, related to this point from the original comment:

>due to [permission constraints](https://help.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token#permissions-for-the-github_token) of GitHub Actions, labels cannot be applied by actions if the originating repository is a fork of Gutenberg.

A commit pushed to a branch originating from a fork would fall under these restrictions.

It works for commits to master, since it's within the source repository, committed by someone with permissions to push to master.

desrosj reviewed (commented):
Thanks for working on this, @aduth! Can you confirm where this prompt will show up? Reading through the code, it looks like it will be added inline as a comment. Do automated comments notify the person like normal issue/pull request activity?

noisysocks commented:
> Can you confirm where this prompt will show up?

It will show up as a comment by `@ghbot` in the pull request.

> Do automated comments notify the person like normal issue/pull request activity?

Yes, it's as if a real human commented on the pull request.

aduth commented:
> The API usage looks correct, however, the API will throw a `404` response when the user isn't known.
> From what I can tell, `got()` will probably consider that an error. I think you need to set the [throwhttperrors](https://www.npmjs.com/package/got#throwhttperrors) option? But that's just from glancing at the docs.

Thanks. Reflecting on this, I think I may refactor to remove the `got` dependency altogether, and to lean on the fact that a 404 status code is returned. This is much simpler to do in the base `https` module, since we don't need to deal with processing the content. The only way the content is used right now is in the debug logging, though I don't think it's strictly necessary.

https://nodejs.org/api/https.html#https_https_get_url_options_callback

Maybe a `HEAD` request would be most appropriate. It appears to work just as well in my testing.

>Please also make sure the request comes with a clear UserAgent that our systems team can differentiate from other random clients to avoid unintentional blocking/rate-limiting.

Will do!

>This is great! I'd use a few contractions -- they'll make it sound a little friendlier/less formal -- and tighten up a sentence here and there, but nothing big:

Thanks for the copy review, @michelleweber ! I'll make these adjustments.

> Can you confirm where this prompt will show up? Reading through the code, it looks like it will be added inline as a comment. Do automated comments notify the person like normal issue/pull request activity?

What @noisysocks said 😄 

I believe it should appear identically as in the bot comment above, https://github.com/WordPress/gutenberg/pull/21221#issuecomment-605341863 .

aduth committed:
Project Management Automation: Fetch profile by https HEAD

aduth committed:
Project Management Automation: Fix hostname to use hostname

aduth committed:
Project Management Automation: Add hasWordPressProfile tests

aduth committed:
Project Management Automation: Add request User-Agent header

aduth committed:
Project Management Automation: Update contributor prompt text

aduth committed:
Project Management Automation: Rename addFirstTimeContributorLabel to firstTimeContributor

aduth committed:
Project Management Automation: Expand function comment to include new behavior

aduth committed:
Project Management Automation: Expand CHANGELOG to include rename

aduth commented:
> Thanks. Reflecting on this, I think I may refactor to remove the `got` dependency altogether, and to lean on the fact that a 404 status code is returned. This is much simpler to do in the base `https` module, since we don't need to deal with processing the content. The only way the content is used right now is in the debug logging, and I don't think it's strictly necessary.
> 
> https://nodejs.org/api/https.html#https_https_get_url_options_callback
> 
> Maybe a `HEAD` request would be most appropriate. It appears to work just as well in my testing.

Updated in eb1dff6 (with additional revisions in 83ef4d69f716d50e12bf728454648f2fd6ce2cf3, 26ddea3bccf87ffad278c5a59503873b2851da05).

>Please also make sure the request comes with a clear UserAgent that our systems team can differentiate from other random clients to avoid unintentional blocking/rate-limiting.

Added in 0d338e5.

I went with `Gutenberg/project-management-automation`. Happy to change if it can be further improved.

>This is great! I'd use a few contractions -- they'll make it sound a little friendlier/less formal -- and tighten up a sentence here and there, but nothing big:

Updated the prompt in 6bedde8.

noisysocks reviewed (approved):
The code looks good to me and Travis CI suggests that your tests and type-checks are passing. There's only one way to see if it actually works 😀🚢

aduth commented:
Let's give it a try!

aduth commented:
And... it works! https://github.com/WordPress/gutenberg/pull/21240#issuecomment-607504789 🎉

noisysocks commented:
Nice! I wonder if it's worth including an @ mention of the PR author in the comment text.

aduth commented:
> Nice! I wonder if it's worth including an @ mention of the PR author in the comment text.

It's a good idea! It might require a small refactoring of the current implementation, since the current text is just a static string, and such a change would need to adjust it to be generated dynamically instead.

aduth commented:
> Nice! I wonder if it's worth including an @ mention of the PR author in the comment text.

See #21384
